### PR TITLE
Signaturehelp cancellation and some caching

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
* SignatureHelp request is now cancellable - I saw this request a bunch in a user's log
* SignatureHelp request cleaned up a bit
* Cache `CmdletInfo`'s and synopsis' of `CmdletInfo`s.
* Renamed `NounExclusionList` to `s_nounExclusionList`

### CODE REVIEW

**Comment by TylerLeonhardt** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs  1
- src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=5323898&bid=17422761)

**Comment by TylerLeonhardt** [comment]
I could add logging when it reads from the cache but then I'd need to pass in the logger.

File: src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs [hunk 1]
```diff
@@ -16,20 +16,26 @@ namespace Microsoft.PowerShell.EditorServices.Services.PowerShellContext
...
-            NounExclusionList.TryAdd("PSRepository", true);
+            s_nounExclusionList.TryAdd("Module", true);
+            s_nounExclusionList.TryAdd("Script", true);
+            s_nounExclusionList.TryAdd("Package", true);
+            s_nounExclusionList.TryAdd("PackageProvider", true);
+            s_nounExclusionList.TryAdd("PackageSource", true);
+            s_nounExclusionList.TryAdd("InstalledModule", true);
+            s_nounExclusionList.TryAdd("InstalledScript", true);
+            s_nounExclusionList.TryAdd("ScriptFileInfo", true);
+            s_nounExclusionList.TryAdd("PSRepository", true);
```
Comment (rkeithhill) [inline src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs:38]: According to the comment on line 62, shouldn't `PowerShellGet` be in this list?

Comment (rkeithhill) [reply src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs:38]: Ah nevermind, this is the noun list, duh.  Not the package name list.

Comment (rkeithhill) [reply src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs:38]: That said, I see more nouns from `PowerShellGet` 2.2.3 than appear in the list above.  `gcm -m PowerShellGet | % Noun | % Sort -uniq`  Is that a problem?

Comment (TylerLeonhardt) [reply src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs:38]: Yeah I just renamed the variable. We should have a better check for verb-noun when PSGet uses the same noun as an inbox module `Module` as an example.

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs [hunk 2]
```diff
@@ -61,28 +66,28 @@ public async Task<SignatureHelp> Handle(SignatureHelpParams request, Cancellatio
...
+                return new SignatureHelp();
+            }
 
-            if (parameterSets != null)
+            SignatureInformation[] signatures = new SignatureInformation[parameterSets.Signatures.Length];
+            for (int i = 0; i < signatures.Length; i++)
             {
-                signatures = new SignatureInformation[parameterSets.Signatures.Length];
-                for (int i = 0; i < signatures.Length; i++)
+                var parameters = new ParameterInformation[parameterSets.Signatures[i].Parameters.Count()];
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs:77]: Not really sure what the best estimate could be here... so I'll just leave it empty.

[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs]: Added!

File: src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs (hunk 1 above)
Comment (TylerLeonhardt) [reply src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs]: I've added a s_cmdletExclusionList now which will handle that ^

**Comment by TylerLeonhardt** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs  1
- src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs  2
         

Clones added
============
- src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=5323898&bid=17422761)

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs [hunk 3]
```diff
@@ -71,15 +70,13 @@ public async Task<SignatureHelp> Handle(SignatureHelpParams request, Cancellatio
                 return new SignatureHelp();
             }
 
-            SignatureInformation[] signatures = new SignatureInformation[parameterSets.Signatures.Length];
+            var signatures = new SignatureInformation[parameterSets.Signatures.Length];
```
Comment (rkeithhill) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs:73]: 👍

**Comment by TylerLeonhardt** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs  1
- src/PowerShellEditorServices/Services/PowerShellContext/Utilities/CommandHelpers.cs  2
         

Clones added
============
- src/PowerShellEditorServices/Services/TextDocument/Handlers/SignatureHelpHandler.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=5323898&bid=17422761)

**Pull request merged**